extensions/iidm/src/InjectionObservabilityXmlSerializer.cpp

```diff
@@ -0,0 +1,86 @@
+}
+
+Extension& InjectionObservabilityXmlSerializer::read(Extendable& extendable, converter::xml::NetworkXmlReaderContext& context) const {
+    bool observable = context.getReader().getOptionalAttributeValue(OBSERVABLE, false);
+
+    InjectionObservabilityAdder adder = extendable.newExtension<InjectionObservabilityAdder>().withObservable(observable);
+
+    context.getReader().readUntilEndElement(getExtensionName(), [&adder, &context]() {
+        if (context.getReader().getLocalName() == QUALITY_P) {
+            const auto& standardDeviation = context.getReader().getAttributeValue<double>(STANDARD_DEVIATION);
+            const auto& redundant = context.getReader().getAttributeValue<bool>(REDUNDANT);
+            adder.withStandardDeviationP(standardDeviation).withRedundantP(redundant);
+        } else if (context.getReader().getLocalName() == QUALITY_Q) {
+            const auto& standardDeviation = context.getReader().getAttributeValue<double>(STANDARD_DEVIATION);
+            const auto& redundant = context.getReader().getAttributeValue<bool>(REDUNDANT);
+            adder.withStandardDeviationQ(standardDeviation).withRedundantQ(redundant);
+        } else if (context.getReader().getLocalName() == QUALITY_V) {
+            const auto& standardDeviation = context.getReader().getAttributeValue<double>(STANDARD_DEVIATION);
+            const auto& redundant = context.getReader().getAttributeValue<bool>(REDUNDANT);
+            adder.withStandardDeviationV(standardDeviation).withRedundantV(redundant);
+        } else {
+            throw PowsyblException(stdcxx::format("Unexpected element: %1%", context.getReader().getLocalName()));
+        }
+    });
+
+    adder.add();
+    return extendable.getExtension<InjectionObservability>();
+}
+
+void InjectionObservabilityXmlSerializer::write(const Extension& extension, converter::xml::NetworkXmlWriterContext& context) const {
+    const auto& injectionObservability = safeCast<InjectionObservability>(extension);
+    context.getWriter().writeOptionalAttribute(OBSERVABLE, injectionObservability.isObservable(), false);
+    writeOptionalQuality(QUALITY_P, injectionObservability.getQualityP(), context.getWriter());
+    writeOptionalQuality(QUALITY_Q, injectionObservability.getQualityQ(), context.getWriter());
+    writeOptionalQuality(QUALITY_V, injectionObservability.getQualityV(), context.getWriter());
+}
+
+void InjectionObservabilityXmlSerializer::writeOptionalQuality(const std::string& elementName, const stdcxx::CReference<ObservabilityQuality>& quality, xml::XmlStreamWriter& writer) const {
+    if (quality) {
+        writer.writeStartElement(getNamespacePrefix(), elementName);
+        writer.writeAttribute(STANDARD_DEVIATION, quality.get().getStandardDeviation());
+        if (quality.get().isRedundant()) {
+            writer.writeAttribute(REDUNDANT, *quality.get().isRedundant());
```

Comment:
From my point of view, it's not equivalent to the Java code:
```
XmlUtil.writeOptionalBoolean(REDUNDANT, quality.isRedundant(), false, writer);
```

If `quality.isRedundant()` is null, there is a NullPointerException due to the automatic conversion from `Boolean` to `boolean`. Moreover, if the value is `false`, the attribute is not serialized.

Reply:
I will check with powsybl-core dev team to know if we really need an `optional<bool>` there

Reply:
https://github.com/powsybl/powsybl-core/pull/1941#pullrequestreview-854464591